### PR TITLE
Add thread statistics

### PR DESCRIPTION
**Pull request by Skelmis**
This PR is a work in progress, however, the code here can be reviewed against.

This PR will bring various thread-related statistics to Nextcord, with Previous acting as a data provider to my statistics service.

Done:
- Catch the initial topic on thread creation via button click
- Catch new thread messages
- Catch thread creation and deletion
- Add further buttons to the help menu to add further topics

TODO:

- Rework the topic command
    - Use a drop-down to provide generic categories (Statistic specific)
    - Use the user-provided input as channel name and specific_topic for statistics

Note ``algoliasearch`` is added as dep as a side effect due to it being required to run master branch.

### CODE REVIEW

File: cogs/help.py [hunk 1]
```diff
@@ -259,6 +296,11 @@ async def on_thread_member_remove(self, member: ThreadMember):
             return
 
         await close_help_thread("EVENT", thread, thread_author)
+        await stats_client.update_thread(
```
Comment (Soheab) [inline cogs/help.py:299]: Why is this only in the member remove/leave event? Shouldn't it be in the close thread func instead?

File: cogs/help.py [hunk 2]
```diff
@@ -135,12 +140,24 @@ def is_allowed(message: Message) -> bool:
...
-            await self.view.bot.wait_for("message", timeout=WAIT_FOR_TIMEOUT, check=is_allowed)
+            m: Message = await self.view.bot.wait_for("message", timeout=WAIT_FOR_TIMEOUT, check=is_allowed)
         except TimeoutError:
             await close_help_thread("TIMEOUT [launch_wait_for_message]", thread, interaction.user)
             return
         else:
             await thread.send(f"<@&{HELPER_ROLE_ID}>", delete_after=5)
+
+            role = thread.guild.get_role(HELPER_ROLE_ID)
+            is_helper = role in m.author.roles
```
Comment (Soheab) [inline cogs/help.py:151]: Could do `is_helper = m.author.get_role(HELPER_ROLE_ID)` instead

File: cogs/help.py [hunk 3]
```diff
@@ -233,20 +255,35 @@ def __init__(self, bot):
         self.close_empty_threads.start()
         self.bot.loop.create_task(self.create_views())
 
+        # Funky, but saves rewriting everything.
+        global stats_client
```
Comment (Soheab) [inline cogs/help.py:259]: Can't it be a class variable instead?

File: cogs/help.py [hunk 4]
```diff
@@ -233,20 +255,35 @@ def __init__(self, bot):
...
-                message.type is MessageType.pins_add:
-            await message.delete(delay = 10)
+
+        if isinstance(message.channel, Thread) and message.channel.parent_id == HELP_CHANNEL_ID:
+            if message.type is MessageType.pins_add:
+                await message.delete(delay=10)
+
+            else:
+                role = message.guild.get_role(HELPER_ROLE_ID)
+                is_helper = role in message.author.roles
```
Comment (Soheab) [inline cogs/help.py:279]: Same as above

**Review by TAG-Epic — COMMENTED** [review]
You also forgot to update the nomad file

File: cogs/utils/stats.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+class StatsClient:
+    """A simple interaction client for https://nextcord.koldfusion.xyz"""
+    def __init__(self, session: ClientSession):
+        self.session: ClientSession = session
+        self.base_url: str = "https://nextcord.koldfusion.xyz/api/v1/"
+
+        api_key: str = os.environ.get("STATS_API_KEY")
+        if not api_key:
+            raise KeyError(
```
[the remark this answers is not in the document]
Comment (Skelmis) [reply cogs/utils/stats.py:20]: I'm not giving everyone access, just whoever is running previous.

File: cogs/help.py (hunk 1 above)
Comment (Skelmis) [reply cogs/help.py:299]: It requires too much context to be moved there without extra refactoring, although it would be ideal.

**Comment by Skelmis** [comment]
Unfortunately due to a combination of a new job and university restarting the feature-set I originally set out to implement is going to require too much work and time that I no longer have.

The server is open-source and can be found [here](https://github.com/Skelmis/nextcord-stats). This PR and the server would be at around 85-90% complete at best guess, testing & fixes not included.

**Pull request closed without merging**